loader_hub/joplin/base.py

```diff
@@ -0,0 +1,105 @@
+        base_url = f"http://{host}:{port}"
+        self._get_note_url = (
+            f"{base_url}/notes?token={access_token}"
+            f"&fields=id,parent_id,title,body,created_time,updated_time&page={{page}}"
+        )
+        self._get_folder_url = (
+            f"{base_url}/folders/{{id}}?token={access_token}&fields=title"
+        )
+        self._get_tag_url = (
+            f"{base_url}/notes/{{id}}/tags?token={access_token}&fields=title"
+        )
+
+    def _get_token_from_env(self) -> str:
+        if "JOPLIN_ACCESS_TOKEN" in os.environ:
+            return os.environ["JOPLIN_ACCESS_TOKEN"]
+        else:
+            raise ValueError(
+                "You need to provide an access token to use the Joplin reader. You may provide it as an argument or set the JOPLIN_ACCESS_TOKEN environment variable."
+            )
+
+    def _get_notes(self) -> Iterator[Document]:
+        has_more = True
+        page = 1
+        while has_more:
+            req_note = urllib.request.Request(self._get_note_url.format(page=page))
+            with urllib.request.urlopen(req_note) as response:
+                json_data = json.loads(response.read().decode())
+                for note in json_data["items"]:
+                    metadata = {
+                        "source": LINK_NOTE_TEMPLATE.format(id=note["id"]),
+                        "folder": self._get_folder(note["parent_id"]),
+                        "tags": self._get_tags(note["id"]),
+                        "title": note["title"],
+                        "created_time": self._convert_date(note["created_time"]),
+                        "updated_time": self._convert_date(note["updated_time"]),
+                    }
+                    yield Document(text=note["body"], extra_info=metadata)
+
+                has_more = json_data["has_more"]
+                page += 1
+
+    def _get_folder(self, folder_id: str) -> str:
+        req_folder = urllib.request.Request(self._get_folder_url.format(id=folder_id))
+        with urllib.request.urlopen(req_folder) as response:
+            json_data = json.loads(response.read().decode())
+            return json_data["title"]
+
+    def _get_tags(self, note_id: str) -> List[str]:
+        req_tag = urllib.request.Request(self._get_tag_url.format(id=note_id))
+        with urllib.request.urlopen(req_tag) as response:
+            json_data = json.loads(response.read().decode())
+            return ','.join([tag["title"] for tag in json_data["items"]])
+
+    def _convert_date(self, date: int) -> str:
+        return datetime.fromtimestamp(date / 1000).strftime("%Y-%m-%d %H:%M:%S")
+
+    def lazy_load(self) -> Iterator[Document]:
+        yield from self._get_notes()
+
+    def load_data(self) -> List[Document]:
+        return list(self.lazy_load())
+
+    def load_langchain_documents(self, **load_kwargs: Any) -> List[LCDocument]:
+        """Load data in LangChain document format."""
+        docs = self.load_data(**load_kwargs)
```

Comment:
don't think you need to override this method

Reply:
right, thanks! saw it in one of the readers I used as reference (ObsidianReader), but now I see that BaseReader already implements it. removing.